api/task_manager.cc

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/coroutine.hh>
+
+#include "task_manager.hh"
+#include "api/api-doc/task_manager.json.hh"
+#include "db/system_keyspace.hh"
+#include "column_family.hh"
+#include "unimplemented.hh"
+#include "storage_service.hh"
+
+#include <utility>
+#include <boost/range/adaptors.hpp>
+
+namespace api {
+
+namespace tm = httpd::task_manager_json;
+using namespace json;
+
+inline bool filter_tasks(task_manager::task_ptr task, std::unordered_map<sstring, sstring>& query_params) {
+    return (!query_params.contains("keyspace") || query_params["keyspace"] == task->get_status().keyspace) &&
+        (!query_params.contains("table") || query_params["table"] == task->get_status().table);
+}
+
+struct full_task_status {
+    task_manager::task::status task_status;
+    task_manager::task::progress progress;
+    std::string module;
+    task_id parent_id;
+    is_abortable abortable;
+};
+
+tm::task_status make_status(full_task_status status) {
+    auto start_time = db_clock::to_time_t(status.task_status.start_time);
+    auto end_time = db_clock::to_time_t(status.task_status.end_time);
+    ::tm st, et;
+    ::gmtime_r(&end_time, &et);
+    ::gmtime_r(&start_time, &st);
+
+    tm::task_status res{};
+    res.id = status.task_status.id.to_sstring();
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = bool(status.abortable);
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.sequence_number = status.task_status.sequence_number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.progress.total;
+    res.progress_completed = status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        co_return coroutine::return_exception(httpd::bad_param_exception("Invalid task id"));
```

Comment:
It's not necessarily invalid, "task not found" would be more appropriate.
We can verify the uuid validuty but that would be an overkill